src/shims/unix/socket.rs

```diff
@@ -22,15 +29,71 @@ impl FileDescription for SocketPair {
     ) -> InterpResult<'tcx, io::Result<()>> {
         Ok(Ok(()))
     }
+
+    fn read<'tcx>(
+        &mut self,
+        _communicate_allowed: bool,
+        bytes: &mut [u8],
+        _tcx: TyCtxt<'tcx>,
+    ) -> InterpResult<'tcx, io::Result<usize>> {
+        let request_byte_size = bytes.len();
+        let data_size = self.readbuf.borrow_mut().len();
+
+        // FIXME: blocking is currently not supported
+        if data_size == 0 {
+            throw_unsup_format!("socketpair read: blocking isn't supported yet");
+        }
```

Comment:
as Ralf mentioned in the corresponding issue

> store during socket creation whether the socket is blocking, and then in read/write, if we would have to block, throw_unsup_format!.

This should only bail out as unsupported for blocking sockets.

Reply:
Currently all sockets are blocking as we're rejecting `type` set to `... | SOCK_NONBLOCK`.

However, @tiif said elsewhere tokio needs non-blocking sockets, and yet the tokio tests still pass despite non-blocking sockets now being rejected, so something is odd.

Reply:
Oh strange, we can remove the `socketpair` shim entirely and our test suite still passes. Looks like that shim got added without *any* test coverage. That would explain the confusion.

Reply:
I think to actually invoke the ``libc::socketpair`` from tokio, we need to use ``UnixStream::pair``, This uses the wrapper function from ``mio`` that calls ``libc::socketpair``. I can add this to the test. 

```rust
use tokio::net::UnixStream;

#[tokio::main]
async fn main() {
    let _a = UnixStream::pair();
}
```
According to strace, the test above invoked 
```
socketpair(AF_UNIX, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0, [6, 7]) = 0
socketpair(AF_UNIX, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0, [9, 10]) = 0
```

But strace also says the test below invoked ``socketpair`` too, so I am not sure why ``socketpair`` has no test coverage
```rust
#[tokio::main]
async fn main() {
}
```
strace output for the test above:
```
socketpair(AF_UNIX, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0, [6, 7]) = 0

```

Reply:
You're adding tests that directly call `socketpair`, that's enough. No need to worry about also hitting this via tokio (yet).

I would assume tokio uses socketpair somewhere in its runtime internally, and that's why it came up... but maybe it doesn't always do that, or so?

Reply:
yea, it doesn't always invoke that. I think it depends on thread timings.

Reply:
Turns out we did have test coverage but I accidentally removed it... oopsie.
Will be fixed by https://github.com/rust-lang/miri/pull/3628.